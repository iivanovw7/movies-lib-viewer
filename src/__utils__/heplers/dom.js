/**
 * Checks if passed content is identical to the content of element passed,
 *  or verifies if element contains value passed in params.
 *
 * @param {node | element} element
 *  Target element, which content is going to be verified.
 * @param {*} value
 *  Comparing value.
 * @param {boolean} includes
 *  If `true` - full element is going to be compared,
 *   if `false` - checks if element content contains target value.
 *
 * @author Igor Ivanov
 */
export function checkContent(element, value, includes = false) {
  // prettier-ignore
  expect( includes
    ? element.text().includes(value)
    : element.text()
  )
    .toBe( includes
      ? true
      : value
    );
}

export default {
  checkContent,
};
