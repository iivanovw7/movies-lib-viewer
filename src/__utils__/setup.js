/**
 * Module contains tests setup
 * @module __utils__/setup
 * @author Igor Ivanov
 */
import 'core-js/stable';
import 'regenerator-runtime/runtime';
import Enzyme, { configure } from 'enzyme';
import jsdom from 'jsdom';

// eslint-disable-next-line import/no-extraneous-dependencies
const Adapter = require('enzyme-adapter-react-16');

Enzyme.configure({ adapter: new Adapter() });

const dom = new jsdom.JSDOM('<!DOCTYPE html><html><body></body></html>', {
  url: 'http://localhost:4428',
  pretendToBeVisual: true,
});
const win = dom.window;

global.document = win.document;
global.window = win;
global.localStorage = win.localStorage;
global.requestAnimationFrame = win.requestAnimationFrame;
global.cancelAnimationFrame = win.cancelAnimationFrame;

// eslint-disable-next-line no-shadow
global.window.IntersectionObserver = class IntersectionObserver {
  // eslint-disable-next-line no-useless-constructor,require-jsdoc,no-empty-function
  constructor() {}

  // eslint-disable-next-line class-methods-use-this,require-jsdoc
  observe(cb) {
    return cb;
  }

  // eslint-disable-next-line class-methods-use-this,require-jsdoc
  unobserve() {
    return null;
  }
};

global.window.IntersectionObserverEntry = {};
global.window.IntersectionObserverEntry.prototype = {
  intersectionRatio: {},
};

configure({ adapter: new Adapter() });
