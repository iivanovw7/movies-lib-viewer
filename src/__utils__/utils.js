/**
 * Module contains test utils
 * @module __utils__/utils
 * @author Igor Ivanov
 */
import chalk from 'chalk';
/* eslint-disable */
export const testName = (title, description) => `[${chalk.yellow(title)}]: ${description}`;
/* eslint-enable */
