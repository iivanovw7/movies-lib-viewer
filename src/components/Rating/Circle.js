/**
 * Module contains rating Circle
 * @module components/Rating/Circle
 * @author Igor Ivanov
 */
import { transparentize } from 'polished';
import styled from 'styled-components';

import { colorSet } from '../../config/styles';

const Circle = styled.circle`
  cx: 19;
  cy: 19;
  fill: ${colorSet.dark.headerBackgroundSecondary};
  r: 15.91549430918954;
  stroke: ${(props) => props.stroke || colorSet.dark.headerBackgroundSecondary};
`;

export default Circle;
