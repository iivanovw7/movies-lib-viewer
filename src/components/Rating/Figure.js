/**
 * Module contains rating figure element
 * @module components/Rating/Figure
 * @author Igor Ivanov
 */
import styled from 'styled-components';

const Figure = styled.figure`
  max-width: ${(props) => props.maxSize};
  ${(props) => props.styling};
  vertical-align: middle;
`;

export default Figure;
