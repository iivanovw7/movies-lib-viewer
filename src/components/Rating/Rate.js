/**
 * Module contains rating rate element
 * @module components/Rating/Rate
 * @author Igor Ivanov
 */
import styled from 'styled-components';

const Rate = styled.g`
  fill: ${(props) => props.textColor};
  transform: translateY(0.25em);
`;

export default Rate;
