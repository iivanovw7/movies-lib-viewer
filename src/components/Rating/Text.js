/**
 * Module contains rating text
 * @module components/Rating/Text
 * @author Igor Ivanov
 */
import styled from 'styled-components';

const Text = styled.text`
  font-size: 0.8em;
  line-height: 1;
  text-anchor: middle;
`;

export default Text;
