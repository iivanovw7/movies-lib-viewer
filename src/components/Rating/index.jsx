/**
 * Module contains carousel rating meter.
 * @module containers/Carousel/Poster
 * @author Igor Ivanov
 */
import * as PropTypes from 'prop-types';
import { multiply, subtract, divide, compose, __, cond, gte, T, always } from 'ramda';
import React from 'react';

import Circle from './Circle';
import Figure from './Figure';
import Rate from './Rate';
import Text from './Text';

const INITIAL_OFFSET = 25;

const colorSet = ['#ff0b0b', '#eac000', '#007b00'];

const Rating = (props) => {
  const {
    value,
    min,
    max,
    maxSize,
    textColor,
    styling,
    strokeColor,
    trailStrokeColor,
    strokeWidth,
    trailStrokeWidth,
    trailSpaced,
  } = props;

  const percentage = compose(Math.abs, subtract(min), multiply(100), divide(__, max - min))(value);
  const getTrailColor = cond([
    [gte(__, 60), always(colorSet[2])],
    [gte(__, 40), always(colorSet[1])],
    [T, always(colorSet[0])],
  ]);

  return (
    <Figure styling={styling} maxSize={maxSize}>
      <svg viewBox="0 0 38 38">
        <Circle
          stroke={trailStrokeColor}
          strokeWidth={trailStrokeWidth}
          strokeDasharray={trailSpaced ? 1 : 0}
        />
        <Circle
          stroke={strokeColor || getTrailColor(percentage)}
          strokeWidth={strokeWidth}
          strokeDasharray={`${percentage} ${100 - percentage}`}
          strokeDashoffset={INITIAL_OFFSET}
        />
        <Rate textColor={textColor}>
          <Text x="50%" y="50%">
            {value}
          </Text>
        </Rate>
      </svg>
    </Figure>
  );
};

Rating.propTypes = {
  value: PropTypes.number.isRequired,
  min: PropTypes.number.isRequired,
  max: PropTypes.number.isRequired,
  maxSize: PropTypes.string,
  textColor: PropTypes.string,
  strokeColor: PropTypes.string,
  strokeWidth: PropTypes.number,
  styling: PropTypes.array,
  trailStrokeWidth: PropTypes.number,
  trailStrokeColor: PropTypes.string,
  trailSpaced: PropTypes.bool,
};

Rating.defaultProps = {
  maxSize: '50px',
  strokeWidth: 4,
  trailStrokeWidth: 4,
  trailStrokeColor: 'transparent',
  trailSpaced: false,
};

export default Rating;
