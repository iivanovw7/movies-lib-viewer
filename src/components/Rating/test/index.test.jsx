/**
 * Module contains Rating component test suite
 */
import { mount, shallow } from 'enzyme';
import React from 'react';
import { css } from 'styled-components';

import { testName } from '../../../__utils__/common';
import { checkContent } from '../../../__utils__/heplers/dom';
import Circle from '../Circle';
import Figure from '../Figure';
import Rating, { colorSet } from '../index';
import Rate from '../Rate';
import Text from '../Text';

const basePropsMock = {
  value: 4,
  min: 0,
  max: 10,
};

// prettier-ignore
describe(testName('Rating', 'component test suite'), function FooterTestSuite() {
  // eslint-disable-next-line require-jsdoc
  const Composition = (props, shouldMount = false) => {
    // eslint-disable-next-line react/jsx-props-no-spreading
    const node = <Rating {...props} />;

    return shouldMount ? mount(node) : shallow(node);
  };

  // eslint-disable-next-line require-jsdoc
  function check(target, name, value) {
    expect(target.prop(name))
      .toEqual(value);
  }

  it('Should match snapshot', () => {
    const component = Composition(basePropsMock);

    expect(component)
      .toMatchSnapshot();
  });

  it('Should render component with all children', () => {
    const component = Composition(basePropsMock);
    const figure = component.find(Figure);
    const text = component.find(Text);

    expect(text)
      .toHaveLength(1);

    check(text, 'x', '50%');
    check(text, 'y', '50%');

    expect(figure)
      .toHaveLength(1);

    check(figure, 'maxSize', '40px');

    expect(component.find(Circle))
      .toHaveLength(2);

    expect(component.find(Rate))
      .toHaveLength(1);

    checkContent(text, String(basePropsMock.value));
  });

  it('Should render component with custom styles', () => {
    const StylesMock = css`
      height: 1.2em;
    `;
    const component = Composition({
      ...basePropsMock,
      styling: StylesMock,
    });

    expect(component.find(Figure).prop('styling'))
      .toBe(StylesMock);
  });

  it('Should verify default strokeColor colors', () => {
    const component = Composition(
      basePropsMock,
      true
    );

    check(
      component.find(Circle).at(1),
      'stroke',
      colorSet[1]
    );

    component.setProps({
      value: 1
    })

    check(
      component.find(Circle).at(1),
      'stroke',
      colorSet[0]
    );

    component.setProps({
      value: 9
    })

    check(
      component.find(Circle).at(1),
      'stroke',
      colorSet[2]
    );
  })

  it('Should render dashed Rating element', () => {
    const component = Composition(basePropsMock);

    check(
      component.find(Circle).at(0),
      'strokeDasharray',
      0
    );

    component.setProps({
      trailSpaced: true
    })

    check(
      component.find(Circle).at(0),
      'strokeDasharray',
      1
    );

  });
});
