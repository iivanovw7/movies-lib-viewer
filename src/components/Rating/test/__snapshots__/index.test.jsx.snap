// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[[33mRating[39m]: component test suite Should match snapshot 1`] = `
<Figure
  maxSize="40px"
>
  <svg
    viewBox="0 0 38 38"
  >
    <Circle
      strokeDasharray={0}
      strokeWidth={3}
    />
    <Circle
      stroke="#eac000"
      strokeDasharray="40 60"
      strokeDashoffset={25}
      strokeWidth={3}
    />
    <Rate>
      <Text
        x="50%"
        y="50%"
      >
        4
      </Text>
    </Rate>
  </svg>
</Figure>
`;
